lib/pymedphys/_dicom/connect/listen.py

```diff
@@ -0,0 +1,207 @@
+        """
+        logging.debug("C-ECHO!")
+        return 0x0000
+
+    def on_association_accepted(self, _):
+        self.association_directory = None
+
+    def on_association_released(self, _):
+
+        if self.on_released_callback:
+            self.on_released_callback(self.association_directory)
+
+    def on_c_store(self, event):
+
+        dataset = event.dataset
+
+        # TODO: Should these be place in a config file?
+        mode_prefixes = {
+            "CT Image Storage": "CT",
+            "Enhanced CT Image Storage": "CTE",
+            "MR Image Storage": "MR",
+            "Enhanced MR Image Storage": "MRE",
+            "Positron Emission Tomography Image Storage": "PT",
+            "Enhanced PET Image Storage": "PTE",
+            "RT Image Storage": "RI",
+            "RT Dose Storage": "RD",
+            "RT Plan Storage": "RP",
+            "RT Structure Set Storage": "RS",
+            "Computed Radiography Image Storage": "CR",
+            "Ultrasound Image Storage": "US",
+            "Enhanced Ultrasound Image Storage": "USE",
+            "X-Ray Angiographic Image Storage": "XA",
+            "Enhanced XA Image Storage": "XAE",
+            "Nuclear Medicine Image Storage": "NM",
+            "Secondary Capture Image Storage": "SC",
+        }
+
+        try:
+            mode_prefix = mode_prefixes[dataset.SOPClassUID.name]
+        except KeyError:
+            mode_prefix = "UN"
+
+        suid = dataset.SeriesInstanceUID
+        series_dir = self.storage_directory / suid
+        series_dir.mkdir(exist_ok=True)
+
+        filename = "{0!s}.{1!s}".format(mode_prefix, dataset.SOPInstanceUID)
+        filepath = series_dir / filename
+
+        if filepath.exists:
+            # TODO Currently just logging this, since I'm unsure of the desired
+            # behaviour if the file has already been recieved.
+            logging.debug("DICOM file already exists, overwriting")
+
+        context = event.context
+        meta = Dataset()
+        meta.MediaStorageSOPClassUID = dataset.SOPClassUID
+        meta.MediaStorageSOPInstanceUID = dataset.SOPInstanceUID
+        meta.ImplementationClassUID = PYNETDICOM_IMPLEMENTATION_UID
+        meta.TransferSyntaxUID = context.transfer_syntax
+
+        # The following is not mandatory, set for convenience
+        meta.ImplementationVersionName = PYNETDICOM_IMPLEMENTATION_VERSION
+        file_ds = FileDataset(
+            filepath.as_posix(), {}, file_meta=meta, preamble=b"\0" * 128
+        )
+        file_ds.update(dataset)
+        file_ds.is_little_endian = context.transfer_syntax.is_little_endian
+        file_ds.is_implicit_VR = context.transfer_syntax.is_implicit_VR
+
+        status_ds = Dataset()
+        status_ds.Status = 0x0000
+
+        try:
+            # We use `write_like_original=False` to ensure that a compliant
+            #   File Meta Information Header is written
+            file_ds.save_as(filepath.as_posix(), write_like_original=False)
```

Comment:
saving a file can take a long time, and create a bottleneck.
Is this something that could be done async ?